application/server/configuration.go

```diff
@@ -219,67 +221,190 @@ func updateSnykOpenBrowserCodeActions(c *config.Config, settings types.Settings)
+	for _, fc := range folderConfigs {
+		incomingMap[fc.FolderPath] = fc
+	}
+	return incomingMap
+}
+
+func gatherAllFolderPaths(incomingMap map[types.FilePath]types.FolderConfig, workspace types.Workspace) map[types.FilePath]bool {
+	allPaths := make(map[types.FilePath]bool)
+
+	// Add incoming paths
+	for path := range incomingMap {
+		allPaths[path] = true
+	}
+
+	// Add stored paths from all workspace folders
+	if workspace != nil {
+		for _, folder := range workspace.Folders() {
+			allPaths[folder.Path()] = true
+		}
+	}
 
-		storedConfig := c.FolderConfig(path)
-		// Never trust the IDE for what the FFs and SAST settings are
+	return allPaths
+}
+
+func processSingleFolderConfig(c *config.Config, path types.FilePath, incomingMap map[types.FilePath]types.FolderConfig, notifier notification.Notifier) (types.FolderConfig, bool) {
+	storedConfig := c.FolderConfig(path)
+	folderConfig := initializeFolderConfig(path, storedConfig)
+
+	// Merge incoming config if present
+	if incoming, hasIncoming := incomingMap[path]; hasIncoming {
+		mergeFolderConfig(&folderConfig, incoming)
+	}
+
+	// Never trust the IDE for what the FFs and SAST settings are - always use stored values
+	if storedConfig != nil {
 		folderConfig.FeatureFlags = storedConfig.FeatureFlags
 		folderConfig.SastSettings = storedConfig.SastSettings
+	}
 
-		// Folder config might be new or changed, so (re)resolve the org before saving it.
-		// We should also check that the folder's org is still valid if the globally set org has changed.
-		// Also, if the config hasn't been migrated yet, we need to perform the initial migration.
-		needsMigration := !storedConfig.OrgMigratedFromGlobalConfig
-		orgSettingsChanged := !folderConfigsOrgSettingsEqual(folderConfig, storedConfig)
-
-		if needsMigration || orgSettingsChanged {
-			updateFolderConfigOrg(c, storedConfig, &folderConfig)
-			err := c.UpdateFolderConfig(&folderConfig) // a lot of methods depend on correct persisted organization values
-			if err != nil {
-				notifier.SendShowMessage(sglsp.MTError, err.Error())
-			}
+	updateFolderOrgIfNeeded(c, storedConfig, &folderConfig, notifier)
+	di.FeatureFlagService().PopulateFolderConfig(&folderConfig)
+
+	configChanged := storedConfig == nil || !cmp.Equal(folderConfig, *storedConfig)
+	if configChanged {
+		err := c.UpdateFolderConfig(&folderConfig)
+		if err != nil {
+			notifier.SendShowMessage(sglsp.MTError, err.Error())
 		}
+	}
 
-		di.FeatureFlagService().PopulateFolderConfig(&folderConfig)
+	return folderConfig, configChanged
+}
 
-		if !cmp.Equal(folderConfig, *storedConfig) {
-			needsToSendUpdateToClient = true
-			err := c.UpdateFolderConfig(&folderConfig)
-			if err != nil {
-				notifier.SendShowMessage(sglsp.MTError, err.Error())
-			}
-		}
+func initializeFolderConfig(path types.FilePath, storedConfig *types.FolderConfig) types.FolderConfig {
+	if storedConfig != nil {
+		return *storedConfig
+	}
+	return types.FolderConfig{FolderPath: path}
+}
```

Comment:
I don't think this is necessary, storedConfig package should never return an empty FolderConfig